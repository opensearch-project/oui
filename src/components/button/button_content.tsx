/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React, { HTMLAttributes, FunctionComponent, Ref } from 'react';
import classNames from 'classnames';
import { CommonProps, keysOf } from '../common';
import { OuiLoadingSpinner } from '../loading';
import { OuiIcon, IconType } from '../icon';

export type ButtonContentIconSide = 'left' | 'right';

const iconSideToClassNameMap: {
  [side in ButtonContentIconSide]: string | null;
} = {
  left: null,
  right: 'ouiButtonContent--iconRight',
};

export const ICON_SIDES = keysOf(iconSideToClassNameMap);

export type OuiButtonContentType = HTMLAttributes<HTMLSpanElement>;

/**
 * *INTERNAL ONLY*
 * This component is simply a helper component for reuse within other button components
 */
export interface OuiButtonContentProps extends CommonProps {
  /**
   * Any `type` accepted by OuiIcon
   */
  iconType?: IconType;
  /**
   * Can only be one side `left` or `right`
   */
  iconSide?: ButtonContentIconSide;
  isLoading?: boolean;
  /**
   * Object of props passed to the <span/> wrapping the content's text/children only (not icon)
   */
  textProps?: HTMLAttributes<HTMLSpanElement> &
    CommonProps & {
      ref?: Ref<HTMLSpanElement>;
      'data-text'?: string;
    };
  iconSize?: 's' | 'm';
}

export const OuiButtonContent: FunctionComponent<
  OuiButtonContentType & OuiButtonContentProps
> = ({
  children,
  textProps,
  isLoading = false,
  iconType,
  iconSize = 'm',
  iconSide = 'left',
  ...contentProps
}) => {
  // Add an icon to the button if one exists.
  let buttonIcon;

  if (isLoading) {
    buttonIcon = (
      <OuiLoadingSpinner className="ouiButtonContent__spinner" size="m" />
    );
  } else if (iconType) {
    buttonIcon = (
      <OuiIcon
        className="ouiButtonContent__icon"
        type={iconType}
        size={iconSize}
        title={textProps?.title}
        color="inherit" // forces the icon to inherit its parent color
      />
    );
  }

  const contentClassNames = classNames(
    'ouiButtonContent',
    iconSide ? iconSideToClassNameMap[iconSide] : null,
    contentProps && contentProps.className
  );

  return (
    <span {...contentProps} className={contentClassNames}>
      {buttonIcon}
      <span {...textProps}>{children}</span>
    </span>
  );
};
