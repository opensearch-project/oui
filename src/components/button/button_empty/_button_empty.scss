/*!
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/**
 * 1. We don't want any of the animations that come inherited from the mixin.
 *    These should act like normal links instead.
 * 2. Change the easing, quickness to not bounce so lighter backgrounds don't flash
 */
.ouiButtonEmpty {
  @include ouiButton;

  border-color: transparent;
  background-color: transparent;
  box-shadow: none;
  // sass-lint:disable-block no-important
  transform: none !important; /* 1 */
  animation: none !important; /* 1 */
  transition-timing-function: ease-in; /* 2 */
  transition-duration: $ouiAnimSpeedFast; /* 2 */
  line-height: inherit;

  .ouiButtonEmpty__content {
    padding: 0 $ouiSizeS;
  }

  .ouiButtonEmpty__text {
    text-overflow: ellipsis;
    overflow: hidden;
  }

  &.ouiButtonEmpty--small {
    height: $ouiButtonHeightSmall;
  }

  &.ouiButtonEmpty--xSmall {
    height: $ouiButtonHeightXSmall;
    font-size: $ouiFontSizeS;
  }

  &:disabled {
    @include ouiButtonContentDisabled;

    color: $ouiButtonColorDisabledText;

    &:focus {
      background-color: transparent;
    }

    &:hover,
    &:focus {
      text-decoration: none;
    }
  }
}


.ouiButtonEmpty--flushLeft,
.ouiButtonEmpty--flushRight,
.ouiButtonEmpty--flushBoth {
  .ouiButtonEmpty__content {
    padding-left: 0;
    padding-right: 0;
  }
}

.ouiButtonEmpty--flushLeft {
  margin-right: $ouiSizeS;
}

.ouiButtonEmpty--flushRight {
  margin-left: $ouiSizeS;
}

// Modifier naming and colors.
$ouiButtonEmptyTypes: (
  primary: $ouiColorPrimaryText,
  danger: $ouiColorDangerText,
  disabled: $ouiButtonColorDisabledText,
  ghost: $ouiColorGhost,
  text: $ouiTextColor,
  success: $ouiColorSecondaryText,
  warning: $ouiColorWarningText,
);

// Create button modifiers based upon the map.
@each $name, $color in $ouiButtonEmptyTypes {
  .ouiButtonEmpty--#{$name} {
    @if ($name == 'ghost') {
      // Ghost is unique and ALWAYS sits against a dark background.
      color: $color;
    } @else if ($name == 'text') {
      // The default color is lighter than the normal text color, make the it the text color
      color: $ouiTextColor;
    } @else {
      // Other colors need to check their contrast against the page background color.
      color: makeHighContrastColor($color, $ouiColorEmptyShade);
    }

    &:focus {
      background-color: transparentize($color, .9);
    }

    &:hover {
      @if ($name == 'disabled') {
        cursor: not-allowed;
      }
    }
  }
}



/* OUI -> EUI Aliases */
$euiButtonEmptyTypes: $ouiButtonEmptyTypes;
/* End of Aliases */
