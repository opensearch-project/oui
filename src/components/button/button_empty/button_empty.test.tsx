/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import { render, screen, act } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { requiredProps } from '../../../test/required_props';

import { OuiButtonEmpty, COLORS, SIZES, FLUSH_TYPES } from './button_empty';
import { ICON_GAPS, ICON_SIDES } from '../button_content';

describe('OuiButtonEmpty', () => {
  test('is rendered', () => {
    const { container } = render(
      <OuiButtonEmpty {...requiredProps}>Content</OuiButtonEmpty>
    );

    expect(container).toMatchSnapshot();
  });

  describe('props', () => {
    describe('isDisabled', () => {
      it('is rendered', () => {
        const { container } = render(<OuiButtonEmpty isDisabled />);

        expect(container).toMatchSnapshot();
      });

      it('renders a button even when href is defined', () => {
        const { container } = render(<OuiButtonEmpty href="#" isDisabled />);

        expect(container).toMatchSnapshot();
      });

      it('renders if passed simply as disabled', () => {
        const { container } = render(<OuiButtonEmpty disabled />);

        expect(container).toMatchSnapshot();
      });
    });

    describe('isLoading', () => {
      it('is rendered', () => {
        const { container } = render(<OuiButtonEmpty isLoading />);

        expect(container).toMatchSnapshot();
      });
    });

    describe('isSelected', () => {
      it('is rendered as true', () => {
        const { container } = render(<OuiButtonEmpty isSelected />);

        expect(container).toMatchSnapshot();
      });

      it('is rendered as false', () => {
        const { container } = render(<OuiButtonEmpty isSelected={false} />);

        expect(container).toMatchSnapshot();
      });
    });

    describe('iconType', () => {
      it('is rendered', () => {
        const { container } = render(<OuiButtonEmpty iconType="user" />);

        expect(container).toMatchSnapshot();
      });
    });

    describe('color', () => {
      COLORS.forEach((color) => {
        test(`${color} is rendered`, () => {
          const { container } = render(<OuiButtonEmpty color={color} />);

          expect(container).toMatchSnapshot();
        });
      });
    });

    describe('size', () => {
      SIZES.forEach((size) => {
        test(`${size} is rendered`, () => {
          const { container } = render(<OuiButtonEmpty size={size} />);

          expect(container).toMatchSnapshot();
        });
      });
    });

    describe('iconSide', () => {
      ICON_SIDES.forEach((iconSide) => {
        test(`${iconSide} is rendered`, () => {
          const { container } = render(
            <OuiButtonEmpty iconType="user" iconSide={iconSide}>
              Content
            </OuiButtonEmpty>
          );

          expect(container).toMatchSnapshot();
        });
      });
    });

    describe('iconGap', () => {
      ICON_GAPS.forEach((iconGap) => {
        test(`${iconGap} is rendered`, () => {
          const { container } = render(
            <OuiButtonEmpty iconType="user" iconGap={iconGap}>
              Content
            </OuiButtonEmpty>
          );

          expect(container).toMatchSnapshot();
        });
      });
    });

    describe('flush', () => {
      FLUSH_TYPES.forEach((flushType) => {
        test(`${flushType} is rendered`, () => {
          const { container } = render(<OuiButtonEmpty flush={flushType} />);

          expect(container).toMatchSnapshot();
        });
      });
    });

    describe('href', () => {
      it('secures the rel attribute when the target is _blank', () => {
        const { container } = render(
          <OuiButtonEmpty href="#" target="_blank" />
        );

        expect(container).toMatchSnapshot();
      });
    });

    describe('onClick', () => {
      it('supports onClick and href', async () => {
        const handler = jest.fn();
        render(<OuiButtonEmpty href="#" onClick={handler} />);

        const user = userEvent.setup();
        await act(async () => {
          await user.click(screen.getByRole('link'));
        });

        expect(handler).toHaveBeenCalledTimes(1);
      });

      it('supports onClick as a button', async () => {
        const handler = jest.fn();
        render(<OuiButtonEmpty onClick={handler} />);

        const user = userEvent.setup();
        await act(async () => {
          await user.click(screen.getByRole('button'));
        });

        expect(handler).toHaveBeenCalledTimes(1);
      });
    });

    test('contentProps is rendered', () => {
      const { container } = render(
        <OuiButtonEmpty contentProps={requiredProps}>Content</OuiButtonEmpty>
      );

      expect(container).toMatchSnapshot();
    });

    test('textProps is rendered', () => {
      const { container } = render(
        <OuiButtonEmpty textProps={requiredProps}>Content</OuiButtonEmpty>
      );

      expect(container).toMatchSnapshot();
    });
  });
});
