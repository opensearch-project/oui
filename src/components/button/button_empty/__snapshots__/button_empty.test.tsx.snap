// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OuiButtonEmpty is rendered 1`] = `
<button
  aria-label="aria-label"
  class="ouiButtonEmpty ouiButtonEmpty--primary testClass1 testClass2"
  data-test-subj="test subject string"
  type="button"
>
  <span
    class="ouiButtonContent ouiButtonEmpty__content"
  >
    <span
      class="ouiButtonEmpty__text"
    >
      Content
    </span>
  </span>
</button>
`;

exports[`OuiButtonEmpty props color danger is rendered 1`] = `
<button
  class="ouiButtonEmpty ouiButtonEmpty--danger"
  type="button"
>
  <span
    class="ouiButtonContent ouiButtonEmpty__content"
  >
    <span
      class="ouiButtonEmpty__text"
    />
  </span>
</button>
`;

exports[`OuiButtonEmpty props color ghost is rendered 1`] = `
<button
  class="ouiButtonEmpty ouiButtonEmpty--ghost"
  type="button"
>
  <span
    class="ouiButtonContent ouiButtonEmpty__content"
  >
    <span
      class="ouiButtonEmpty__text"
    />
  </span>
</button>
`;

exports[`OuiButtonEmpty props color primary is rendered 1`] = `
<button
  class="ouiButtonEmpty ouiButtonEmpty--primary"
  type="button"
>
  <span
    class="ouiButtonContent ouiButtonEmpty__content"
  >
    <span
      class="ouiButtonEmpty__text"
    />
  </span>
</button>
`;

exports[`OuiButtonEmpty props color success is rendered 1`] = `
<button
  class="ouiButtonEmpty ouiButtonEmpty--success"
  type="button"
>
  <span
    class="ouiButtonContent ouiButtonEmpty__content"
  >
    <span
      class="ouiButtonEmpty__text"
    />
  </span>
</button>
`;

exports[`OuiButtonEmpty props color text is rendered 1`] = `
<button
  class="ouiButtonEmpty ouiButtonEmpty--text"
  type="button"
>
  <span
    class="ouiButtonContent ouiButtonEmpty__content"
  >
    <span
      class="ouiButtonEmpty__text"
    />
  </span>
</button>
`;

exports[`OuiButtonEmpty props color warning is rendered 1`] = `
<button
  class="ouiButtonEmpty ouiButtonEmpty--warning"
  type="button"
>
  <span
    class="ouiButtonContent ouiButtonEmpty__content"
  >
    <span
      class="ouiButtonEmpty__text"
    />
  </span>
</button>
`;

exports[`OuiButtonEmpty props contentProps is rendered 1`] = `
<button
  class="ouiButtonEmpty ouiButtonEmpty--primary"
  type="button"
>
  <span
    aria-label="aria-label"
    class="ouiButtonContent ouiButtonEmpty__content testClass1 testClass2"
    data-test-subj="test subject string"
  >
    <span
      class="ouiButtonEmpty__text"
    >
      Content
    </span>
  </span>
</button>
`;

exports[`OuiButtonEmpty props flush both is rendered 1`] = `
<button
  class="ouiButtonEmpty ouiButtonEmpty--primary ouiButtonEmpty--flushBoth"
  type="button"
>
  <span
    class="ouiButtonContent ouiButtonEmpty__content"
  >
    <span
      class="ouiButtonEmpty__text"
    />
  </span>
</button>
`;

exports[`OuiButtonEmpty props flush left is rendered 1`] = `
<button
  class="ouiButtonEmpty ouiButtonEmpty--primary ouiButtonEmpty--flushLeft"
  type="button"
>
  <span
    class="ouiButtonContent ouiButtonEmpty__content"
  >
    <span
      class="ouiButtonEmpty__text"
    />
  </span>
</button>
`;

exports[`OuiButtonEmpty props flush right is rendered 1`] = `
<button
  class="ouiButtonEmpty ouiButtonEmpty--primary ouiButtonEmpty--flushRight"
  type="button"
>
  <span
    class="ouiButtonContent ouiButtonEmpty__content"
  >
    <span
      class="ouiButtonEmpty__text"
    />
  </span>
</button>
`;

exports[`OuiButtonEmpty props href secures the rel attribute when the target is _blank 1`] = `
<a
  class="ouiButtonEmpty ouiButtonEmpty--primary"
  href="#"
  rel="noopener noreferrer"
  target="_blank"
>
  <span
    class="ouiButtonContent ouiButtonEmpty__content"
  >
    <span
      class="ouiButtonEmpty__text"
    />
  </span>
</a>
`;

exports[`OuiButtonEmpty props iconSide left is rendered 1`] = `
<button
  class="ouiButtonEmpty ouiButtonEmpty--primary"
  type="button"
>
  <span
    class="ouiButtonContent ouiButtonEmpty__content"
  >
    <span
      class="ouiButtonContent__icon"
      color="inherit"
      data-ouiicon-type="user"
    />
    <span
      class="ouiButtonEmpty__text"
    >
      Content
    </span>
  </span>
</button>
`;

exports[`OuiButtonEmpty props iconSide right is rendered 1`] = `
<button
  class="ouiButtonEmpty ouiButtonEmpty--primary"
  type="button"
>
  <span
    class="ouiButtonContent ouiButtonContent--iconRight ouiButtonEmpty__content"
  >
    <span
      class="ouiButtonContent__icon"
      color="inherit"
      data-ouiicon-type="user"
    />
    <span
      class="ouiButtonEmpty__text"
    >
      Content
    </span>
  </span>
</button>
`;

exports[`OuiButtonEmpty props iconType is rendered 1`] = `
<button
  class="ouiButtonEmpty ouiButtonEmpty--primary"
  type="button"
>
  <span
    class="ouiButtonContent ouiButtonEmpty__content"
  >
    <span
      class="ouiButtonContent__icon"
      color="inherit"
      data-ouiicon-type="user"
    />
    <span
      class="ouiButtonEmpty__text"
    />
  </span>
</button>
`;

exports[`OuiButtonEmpty props isDisabled is rendered 1`] = `
<button
  class="ouiButtonEmpty ouiButtonEmpty--primary ouiButtonEmpty-isDisabled"
  disabled=""
  type="button"
>
  <span
    class="ouiButtonContent ouiButtonEmpty__content"
  >
    <span
      class="ouiButtonEmpty__text"
    />
  </span>
</button>
`;

exports[`OuiButtonEmpty props isDisabled renders a button even when href is defined 1`] = `
<button
  class="ouiButtonEmpty ouiButtonEmpty--primary ouiButtonEmpty-isDisabled"
  disabled=""
  type="button"
>
  <span
    class="ouiButtonContent ouiButtonEmpty__content"
  >
    <span
      class="ouiButtonEmpty__text"
    />
  </span>
</button>
`;

exports[`OuiButtonEmpty props isDisabled renders if passed simply as disabled 1`] = `
<button
  class="ouiButtonEmpty ouiButtonEmpty--primary ouiButtonEmpty-isDisabled"
  disabled=""
  type="button"
>
  <span
    class="ouiButtonContent ouiButtonEmpty__content"
  >
    <span
      class="ouiButtonEmpty__text"
    />
  </span>
</button>
`;

exports[`OuiButtonEmpty props isLoading is rendered 1`] = `
<button
  class="ouiButtonEmpty ouiButtonEmpty--primary ouiButtonEmpty-isDisabled"
  disabled=""
  type="button"
>
  <span
    class="ouiButtonContent ouiButtonEmpty__content"
  >
    <span
      class="ouiLoadingSpinner ouiLoadingSpinner--medium ouiButtonContent__spinner"
    />
    <span
      class="ouiButtonEmpty__text"
    />
  </span>
</button>
`;

exports[`OuiButtonEmpty props isSelected is rendered as false 1`] = `
<button
  aria-pressed="false"
  class="ouiButtonEmpty ouiButtonEmpty--primary"
  type="button"
>
  <span
    class="ouiButtonContent ouiButtonEmpty__content"
  >
    <span
      class="ouiButtonEmpty__text"
    />
  </span>
</button>
`;

exports[`OuiButtonEmpty props isSelected is rendered as true 1`] = `
<button
  aria-pressed="true"
  class="ouiButtonEmpty ouiButtonEmpty--primary"
  type="button"
>
  <span
    class="ouiButtonContent ouiButtonEmpty__content"
  >
    <span
      class="ouiButtonEmpty__text"
    />
  </span>
</button>
`;

exports[`OuiButtonEmpty props size l is rendered 1`] = `
<button
  class="ouiButtonEmpty ouiButtonEmpty--primary ouiButtonEmpty--large"
  type="button"
>
  <span
    class="ouiButtonContent ouiButtonEmpty__content"
  >
    <span
      class="ouiButtonEmpty__text"
    />
  </span>
</button>
`;

exports[`OuiButtonEmpty props size s is rendered 1`] = `
<button
  class="ouiButtonEmpty ouiButtonEmpty--primary ouiButtonEmpty--small"
  type="button"
>
  <span
    class="ouiButtonContent ouiButtonEmpty__content"
  >
    <span
      class="ouiButtonEmpty__text"
    />
  </span>
</button>
`;

exports[`OuiButtonEmpty props size xs is rendered 1`] = `
<button
  class="ouiButtonEmpty ouiButtonEmpty--primary ouiButtonEmpty--xSmall"
  type="button"
>
  <span
    class="ouiButtonContent ouiButtonEmpty__content"
  >
    <span
      class="ouiButtonEmpty__text"
    />
  </span>
</button>
`;

exports[`OuiButtonEmpty props textProps is rendered 1`] = `
<button
  class="ouiButtonEmpty ouiButtonEmpty--primary"
  type="button"
>
  <span
    class="ouiButtonContent ouiButtonEmpty__content"
  >
    <span
      aria-label="aria-label"
      class="ouiButtonEmpty__text testClass1 testClass2"
      data-test-subj="test subject string"
    >
      Content
    </span>
  </span>
</button>
`;
